Sources/IO/Core/HttpDataSetReader/index.js

```diff
@@ -145,26 +147,66 @@ function vtkHttpDataSetReader(publicAPI, model) {
   // Internal method to fetch Array
   function fetchArray(array, options = {}) {
     const arrayId = `${array.ref.id}|${array.vtkClass}`;
-    if (!cachedArrays[arrayId]) {
+
+    // Check if array is present in cache
+    if (model.arrayCachingEnabled && cachedArrays[arrayId]) {
+      // Update last access for cache retention rules
+      cachedArrays[arrayId].lastAccess = new Date();
+      return cachedArrays[arrayId].array;
+    }
+
+    if (!cachedPromises[arrayId]) {
       // Cache the promise while fetching
-      cachedArrays[arrayId] = model.dataAccessHelper
+      cachedPromises[arrayId] = model.dataAccessHelper
         .fetchArray(publicAPI, model.baseURL, array, options)
         .then((newArray) => {
-          // Replace the promise with the array in cache once downloaded
-          cachedArrays[arrayId] = newArray;
+          // Remove cached promise
+          delete cachedPromises[arrayId];
+
+          // Return here if caching is disabled
+          if (!model.arrayCachingEnabled) {
+            return newArray;
+          }
+
+          // sort old cache without new entry (potentially needed for creating space)
+          const sortedArrayCache = Object.entries(cachedArrays).sort((a, b) =>
+            Math.sign(a[1].lastAccess - b[1].lastAccess)
+          );
+
+          // Add array to cache
+          cachedArrays[arrayId] = { array: newArray, lastAccess: new Date() };
+
+          // Check cache size
+          const cacheSizeLimit = model.maxCacheSize * MiB;
+          let cacheSize = Object.values(cachedArrays).reduce(
+            (accSize, entry) => accSize + entry.array.values.byteLength,
+            0
+          );
+
+          // Delete cache entries until size is below the limit
+          while (cacheSize > cacheSizeLimit && sortedArrayCache.length > 0) {
+            const [oldId, entry] = sortedArrayCache.pop();
+            delete cachedArrays[oldId];
+            cacheSize -= entry.array.values.byteLength;
+          }
+
+          // Edge case: If the new entry is bigger than the cache limit
+          if (cacheSize > cacheSizeLimit) {
+            delete cachedArrays[arrayId];
+          }
```

Comment:
You can drop this edge case code handling with some code rearrangements.

```
// Add array to cache
...

// sort cache. Last element is the new array.
```

Additionally, I think we should have a warning if the new entry is bigger than the cache limit.

```
if (!cachedArrays[arrayId]) vtkWarningMacro('Cache size is too small for the dataset')
```

Reply:
Changes implemented as requested.